cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -119,29 +120,30 @@ public PluginEndpoint getPluginEndpoint(NamespaceId namespace,
 
     ArtifactDetail artifactDetail = artifactRepository.getArtifact(artifactId);
     return getPluginEndpoint(namespace, artifactDetail, pluginType, pluginName,
-                             pickParentArtifact(artifactDetail, artifactId), methodName);
+                             getParentArtifacts(artifactDetail, artifactId), methodName);
   }
 
-  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail, Id.Artifact artifact)
+  private List<ArtifactDescriptor> getParentArtifacts(ArtifactDetail artifactDetail, Id.Artifact artifact)
     throws Exception {
-
     // get parent artifacts
     Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
-    if (parentArtifactRanges.isEmpty()) {
-      throw new ArtifactNotFoundException(artifact.toEntityId());
-    }
 
     // just pick the first parent artifact from the set.
-    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+    List<ArtifactDescriptor> artifactDescriptors = new ArrayList<>();
+    for (ArtifactRange artifactRange : parentArtifactRanges) {
+      List<ArtifactDetail> artifactDetails = artifactRepository.getArtifacts(artifactRange);
+      if (!artifactDetails.isEmpty()) {
+        // add the latest version
```

Comment:
this can also cause the same types of problems as before, where the plugin extends an older version but not the latest version. I assume this is being done to limit the number of lookups we have to do through the ArtifactStore.

I think what we can do is add a getPluginClasses() method to ArtifactStore that takes an ArtifactRange instead of an ArtifactId for the parent. Looking for a range of parents amounts to reading the same amount of data from the table but just filtering less out, which should make it as efficient as the existing call.